model/src/main/java/com/chiralbehaviors/CoRE/Ruleform.java

```diff
@@ -104,6 +153,199 @@
         return entity;
     }
 
+    @SuppressWarnings("unchecked")
+    public static <T extends Ruleform> T sanitize(T ruleform) throws IOException {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        serialize(ruleform, os);
+        ByteArrayInputStream is = new ByteArrayInputStream(os.toByteArray());
+        return (T) deserialize(is, ruleform.getClass());
+    }
+
+    public static void serialize(Ruleform ruleform,
+                                 OutputStream os) throws IOException {
+        ObjectMapper objMapper = new ObjectMapper();
+        objMapper.registerModule(new CoREModule());
+        Hibernate4Module module = new Hibernate4Module();
+        module.enable(Feature.FORCE_LAZY_LOADING);
+        objMapper.registerModule(module);
+        objMapper.writerWithDefaultPrettyPrinter()
+                 .writeValue(os, ruleform);
+    }
+
+    public static Map<UUID, Ruleform> slice(Ruleform ruleform,
+                                            Predicate<Ruleform> systemDefinition,
+                                            Map<UUID, Ruleform> sliced,
+                                            Set<UUID> traversed) {
+        map(ruleform, systemDefinition, sliced, traversed);
+        return sliced;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T extends Ruleform> T smartMerge(EntityManager em,
+                                                    T ruleform,
+                                                    Map<UUID, Ruleform> mapped) {
+        return (T) map(em, ruleform, mapped);
+    }
+
+    public static void traverse(Ruleform ruleform,
+                                BiConsumer<Ruleform, Field> traverser) {
+        for (Field field : getInheritedFields(ruleform.getClass())) {
+            if (field.getAnnotation(JoinColumn.class) == null) {
+                continue;
+            }
+            try {
+                field.setAccessible(true);
+                Ruleform value = (Ruleform) field.get(ruleform);
+                if (value != null && !ruleform.id.equals(value.id)) {
+                    traverser.accept(ruleform, field);
+                }
+            } catch (IllegalAccessException e) {
+                throw new IllegalStateException(String.format("IllegalAccess access foreign key field: %s",
+                                                              field),
+                                                e);
+            } catch (IllegalArgumentException e) {
+                throw new IllegalStateException(String.format("Illegal mapped value for field: %s",
+                                                              field),
+                                                e);
+            }
+        }
+    }
+
+    private static Ruleform asFrontier(Ruleform ruleform) {
```

Comment:
I now use a clone of the ruleform as the representation of an exit in the workspace snapshot.  these are ruleforms that are referenced by ruleforms in the workspace.  These are stand ins for ruleforms not in the workspace.  All their FK fields are nulled.  Had to do this because we use Attribute.valueType to determine what kind of an attribute we use.  There's some others, too.  Was hard sweat coming up with this.